pkg/catalog/manager/manager.go

```diff
@@ -242,7 +242,13 @@ func (m *Manager) ValidateKubeVersion(template *v3.CatalogTemplateVersion, clust
 	return nil
 }
 
-func (m *Manager) ValidateRancherVersion(template *v3.CatalogTemplateVersion) error {
+func (m *Manager) ValidateRancherVersion(template *v3.CatalogTemplateVersion, currentAppVersion string) error {
```

Comment:
Knowing the current version of the chart does not change the fact that this chart is not valid for this version of Rancher; rather, it only indicates that this chart is still **compatible** with your current Rancher setup to allow it to be upgraded as we simply **bypass** the rancher version check in this case.

Therefore, I don't think we should be modifying the definition of `ValidateRancherVersion`; IMO this condition should be relocated to `ValidateChartCompatibility`. ​

This would also allow us to remove the changes from `pkg/api/norman/customization/catalog/store.go` and `pkg/api/norman/customization/multiclusterapp/multi_cluster_app.go`; for `pkg/api/norman/customization/multiclusterapp/store.go`, I think we should introduce this bypass logic directly into code instead.

---

Optionally, you could turn this into a helper function like `func (m *Manager) IsUpgrade(template *v3.CatalogTemplateVersion, currentAppVersion string)` and call it here and in `pkg/api/norman/customization/multiclusterapp/store.go` to avoid redundant code.

Reply:
@aiyengar2 I think have this logic in ValidateRancherVersion makes sense because place are bypassing ValidateChartCompatibility and just using ValidateRancherVersion. I also thing the changes in the store and mcapp are desirable. Mcapp is subject to these same things. It is possible an app would be updated using the API/API UI and in this case we still want this validated, that is why we should leave it in the store. We could use a pointer but I'm not sure if it is wise to use pointers where they are no necessary as it invites another point of failure. An empty string indicating that it isn't being used is not much different than using nil.

Reply:
re:
> @aiyengar2 I think have this logic in ValidateRancherVersion makes sense because place are bypassing ValidateChartCompatibility and just using ValidateRancherVersion.

Any part of the code that is bypassing `ValidateChartCompatibility` in favor of directly calling `ValidateRancherVersion` is effectively running its own custom compatibility check logic (e.g. mcapps). 

And I still think that checking whether the template would trigger an app upgrade (and therefore choosing to bypass the Rancher version check) is independent of checking whether a template is valid in a given Rancher version. Whether it would trigger an app upgrade and require additional checks (such as a Rancher version check) is a concern of whatever is executing the compatibility check, whether it be `ValidateChartCompatibility` or a custom compatibility check directly introduced within the code (like mcapps does).

To facilitate areas in the code like mcapps  that are using custom logic instead of the standard `ValidateChartCompatibility`, I think we should introduce `IsUpgrade(template *v3.CatalogTemplateVersion, currentAppVersion string) bool` to `CatalogManager` that has this check so that mcapps could use it as:

```go
if s.catalogManager.IsUpgrade(template, mcappCurrentTemplateVersion) {
    if err := s.catalogManager.ValidateRancherVersion(template, mcappCurrentTemplateVersion); err != nil {
	return httperror.NewAPIError(httperror.InvalidBodyContent, err.Error())
    }
}
```

I believe this should address the concern in:

> It is possible an app would be updated using the API/API UI and in this case we still want this validated, that is why we should leave it in the store.

This could also be how we call it in `ValidateChartCompatibility` as well.

---

re:
> I also thing the changes in the store and mcapp are desirable.

Maybe I'm missing something, but why would this be desirable?

---

Reply:
discussed offline and decided to stick with the current approach on the PR